### ISSUE TITLE
FYI - Detecting Canvas Defender and recovering the original canvas value

### ISSUE THREAD

**Issue by Just-me-ghacks** [ghacksuserjs/ghacks-user.js#458]
[Evaluating the privacy implications of a canvas fingerprinting countermeasure](https://antoinevastel.com/tracking/2018/07/01/eval-canvasdef.html)

> In this post, we showed that contrary to what’s claimed in their article, Canvas Defender can easily be detected. We showed that detecting if a user has Canvas Defender installed can be checked by looking at the string representation of toDataURL. It can also be detected using the MutationObserver API. The advantage of this approach is that it enables to extract the noise vector used by Canvas Defender. Depending on the settings chosen by the user in the extension, this noise vector may remain the same until the user deliberately changes it. Finally, we showed how to recover the original canvas value, even in the presence of Canvas Defender.

Can mozilla and google recover the original canvas value if you use **privacy.resistFingerprinting** to spoof it?

**Comment by Thorin-Oakenpants** [comment]
Just as well we don't recommend canvas defender - personally I think it's premise is flawed. Haven't read the article yet

note: https://kkapsner.github.io/CanvasBlocker/test/detectionTest.html

I don't think anyone cares if their canvas is detected as fake - eg RFP's hash gives it away as a 10x10 white canvas. But ID'ing the actual extension by name could raise entropy. For RFP users, we're already in the RFP set.

> Can mozilla and google recover the original canvas value

Good question. I have no idea.

For Canvas Blocker - @kkapsner
For RFP - @tomrittervg , @arthuredelstein

**Comment by Thorin-Oakenpants** [comment]
From Arthur

> I think RFP is safe from this attack because, when RFP is enabled, the canvas image extraction APIs (such as toDataURL) are sanitized in native code.

**Comment by earthlng** [comment]
Holy shit! Thanks for sharing that link.

This also makes some of our userscripts easily detectable btw ...
Ctrl+K then enter `history.__lookupGetter__('length').toString()` to see the custom function that spoofs history.length

**Comment by kkapsner** [comment]
I think all the points in the article do not apply to CanvasBlocker. The point is that Firefox has some additional WebExtension APIs that allow addon-code to run in the (security) context of the web page (exportFunction). But if you create a WebExtension that also has to run on Chrome or even Edge you cannot use them.

I think I also did a test where I could extract the canvas fingerprint with Canvas Defender installed without any additional code - the addon code was just not loaded fast enough

@earthlng: these APIs would also allow the history.length to be spoofed in a sneaky fashion. To be honest: it would be really simple to include this in CanvasBlocker... ;)

**Comment by kkapsner** [comment]
Just realized they removed the exportFunction documentation: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Content_scripts$compare?locale=en-US&to=1348378&from=1348018
Hopefully they do not want to ditch it.

**Comment by kkapsner** [comment]
Here my test page for the extraction: https://canvasblocker.kkapsner.de/test/settingsLoading.php

Canvas Defender: 
![grafik](https://user-images.githubusercontent.com/1105157/42245505-588de8ea-7f19-11e8-954e-810a25e001ff.png)

CanvasBlocker: 
![grafik](https://user-images.githubusercontent.com/1105157/42245533-6b8dcc3a-7f19-11e8-82f4-8441a59adbff.png) notice the fingerprint icon in the address bar (here I can give you the real data as they are faked ;))

**Comment by Thorin-Oakenpants** [comment]
> Ctrl+K then enter `history.__lookupGetter__('length').toString()` to see the custom function that spoofs history.length

`Ctrl-SHIFT-K` btw. Holy cow. This is all a bit over my head. Does this mean that (aside from kkapsner's info on exportFunction which is a concern if it is removed), that crafty scripts can enumerate unique function names and procure a list of web extensions? That would be a high entropy FP

@kkapsner Wow dude, thanks. I have the utmost faith in you - you've proven it time and time again :heart:

**Comment by Thorin-Oakenpants** [comment]
RFP
```
fingerprint consistent (e5d9fd78536844cc8a4144ddb7a03eb9628f12c7c8b7828f942cadf6efb79ac0) -> good!
```

what on earth is `e5d...` - that's not the RFP 10x10 `a273d6a847f0e2a57fa0161158f12fed`

**Comment by Thorin-Oakenpants** [comment]
Am concerned about that exportFunction being removed. Found your ticket from a while ago ( https://bugzilla.mozilla.org/show_bug.cgi?id=1081385 ) .. component XPConnect, fixed in FF59. Doesn't seem like something they would deprecate. Maybe we should ask Boris

**Comment by Thorin-Oakenpants** [comment]
FYI @antoinevastel

**Comment by Thorin-Oakenpants** [comment]
@antoinevastel et al - https://www.ghacks.net/2018/07/04/study-on-the-effectiveness-of-fingerprinting-countermeasures/ 

PS: I bookmarked and read that paper a couple of days ago :+1: after reading this: https://freedom-to-tinker.com/2018/06/29/against-privacy-defeatism-why-browsers-can-still-stop-fingerprinting/

**Comment by kkapsner** [comment]
Regarding ```e5d...```: I get this as well when disabling CB and using RFP - so it's the correct value. I use a different canvas size (220px × 30px) and therefore the hash is different (also I do not know which hashing algorithm panopticlick is using - I use sha256 of the data-URL).

**Comment by earthlng** [comment]
> Just realized they removed the exportFunction documentation

they moved it to https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Sharing_objects_with_page_scripts

**Comment by earthlng** [comment]
@kkapsner 

> these APIs would also allow the history.length to be spoofed in a sneaky fashion.

can you show me how, please? I can't wrap my head around how to exportFunction a getter in a way that still makes it possible to check the original history.length (to know what to return) while at the same time prevent the page from seeing the custom function.

**Comment by Thorin-Oakenpants** [comment]
I'm just going to @gorhill here becuz 1. it's really interesting that without exportFunction, JS can extract custom function names (I assume I got that right), and 2. i want piece of mind from the owner that uM, uBO are OK

Imagine grabbing the top 200 web extensions in AMO and working out which ones you could ID using this method (see link in OP), eg a unique function name - lets say you could easily ID 150 of them - that's a lot of entropy - unless I've completely missed the point (ELI5 if I have)

**Comment by kkapsner** [comment]
@earthlng: thanks for the new link - good to know that it is still documented.
For the "how to exportFunction a getter" you can check CB-Code as that is exactly what I do: https://github.com/kkapsner/CanvasBlocker/blob/master/lib/intercept.js#L179
You first store the original function (has to be slighly different in your case as you have to store the getter) and then you redefine the getter.

@Thorin-Oakenpants: I do not think that the situation is that bad. Most dddons do not need to provide functions that can be called by the website. But I have not checked and unfortunately all the anti-fingerprinting and some of the privacy addons have to. So if they are not carefully they might make things wore.

**Comment by Thorin-Oakenpants** [comment]
>  But I have not checked and unfortunately all the anti-fingerprinting and some of the privacy addons have to. So if they are not carefully they might make things worse.

Thanks. OK, I'll reword it then ... Imagine grabbing the top 50 "privacy / anti-FP'ing" web extensions in AMO and working out which ones you could ID using this method (see link in OP), eg a unique function name - lets say you could easily ID 30 of them - that's *still* a lot of entropy

I think this might require us to vet all our recommended addons, but I have no idea how to do that

**Comment by kkapsner** [comment]
> I think this might require us to vet all our recommended addons, but I have no idea how to do that
One by one on an individual basis. Basically look in the code of the addon or the "bad" guys and see what of the browsers APIs they are using/changing.

**Comment by Thorin-Oakenpants** [comment]
might as well close this. We don't recommend Canvas Defender ... we :heart: CB .. and CB is not affected cuz kkapsner has his shit together :1st_place_medal: :+1:

**Comment by Thorin-Oakenpants** [comment]
> One by one on an individual basis. Basically look in the code of the addon or the "bad" guys and see what of the browsers APIs they are using/changing.

Yeah, but I can't read that - ~~it's all Greek to me (actually, that's a terrible saying to use, because I actually took classical Greek at university)~~ it's all Hungarian to me
